### ISSUE TITLE
"Non Contiguous Layout" discussion

### ISSUE THREAD

**Issue by gingerbeardman** [glushchenko/fsnotes#1097]
**Describe your feature request**

Labelling of "Non Contiguous Layout" is not clear.

But to decide on better labelling we need a more full understanding of what this option actually does.

It's my understanding:

- disabled
  - macOS will render the layout from top to bottom
  - rendering becomes gradually slower as the document size and complexity increases
  - this is the historic FSNotes behaviour
- enabled
  - macOS will render the layout in chunks, in whatever order it decides
  - rendering of large/complex docs *may* speed up
  - rendering and editing issues *may* crop up

**Questions**

- Given that FSNotes are rendered using a WebView, are we OK to assume that the WebView will take care of correct handling of `setAllowsNonContiguousLayout`?
- Can/should we benchmark performance?
- Can/should we do some testing to see if things break?

**References**
- SubEthaEdit
  - disabled it at one point due to issues https://github.com/subethaedit/SubEthaEdit/issues/103
  - current enabled for more than files of length higher than 4096 (characters?) https://github.com/subethaedit/SubEthaEdit/commit/68b4d01ad0fde9c0554958607c41847a824f176b

**Comment by pdanford** [comment]
Here are my observations using FSNotes on macOS 11.2.3:

1. My work flow is markdown notes of typically of 1000 lines (the largest 25 pages) and with several code fence blocks scattered about.

2. I don't see any slowdown at all with "Non Contiguous Layout" unchecked. It's lightening fast for switching between notes as well as editing as long as Highlightr's keyword highlighting is disabled too [1].

3. For people that have giant notes that benefit from "Non Contiguous Layout" being selected, I would remind them that it's FSNotes, not FSBooks :) I'm guessing that the vast majority of your users are not in the FSBooks group and there is a real perceptual cost by having this option selected; see 6 below.

4. For "Non Contiguous Layout" checked and Highlightr enabled, editing notes with code fences becomes quite difficult between the screen jumping around and slowness between key presses involved in opening/closing code blocks. 

5. IMHO, Highlightr should be removed, but keep the option for background color on fenced code blocks (I believe that part is done outside of Highlightr, right?). Highlightr kills performance during editing and also initial note loads, like 2-4 seconds just for a note switch on a high-end machine. Even initial FSNotes launch suffers when Highlitr is enabled. The code highlighting works fine/much faster in Preview Mode and that's the only place I expect it anyway. That being said, I do like the background highlighting of code blocks in Edit Mode because it helps with logical separation and doesn't have any performance cost that I notice.

6. Highlightr and Non Contiguous Layout can be disabled in Editor options, but I'm guessing many will not know to do it. Then they will have a bad editing experience and conclude FSNotes is too buggy and give up, which is a real shame [2]. I have colleagues that tried FSNotes and decided it was too buggy. Even being told the "buggyness" can be fixed by turning two options off didn't work because they already had that first impression.

7. Keeping "Non Contiguous Layout" and waiting for Apple to fix it probably won't work. I bet they never will.


[1] I disable Highlightr by doing:
```
mv /Applications/FSNotes.app/Contents/Frameworks/Highlightr.framework/Versions/A/Resources/highlight.min.js \
   /Applications/FSNotes.app/Contents/Frameworks/Highlightr.framework/Versions/A/Resources/highlight.min.js.disabled
```
This way I still get the code block background highlighting (because it's part of the "Code block live highlighting (app restart required)" option) in Edit Mode.

[2] FSNotes is a really nice piece of work. Very cool. Thank you!

**Comment by quinncomendant** [comment]
My opinion is that if it is difficult to find an understandable label for this option, it's probably a feature that should not be exposed to the user. 

FSNotes should be able to open and render a large text file quickly and all-at-once without the user needing to choose between two difficult-to-understand compromises.

**Comment by gingerbeardman** [comment]
I agree, I think we should make it automatic in the same way SubEthaEdit did.

**Comment by pdanford** [comment]
But any time "Non Contiguous Layout" is selected, FSNote will appear to have buggy editing for many people (the fault of the Apple library, not the FSNotes code). So IMHO, it is better to have FSnotes slow down loading for some people who use it for large notes rather than for it to appear to have buggy editing (i.e. so never use "Non Contiguous Layout").

As for the slowness loading in Edit Mode, I do not see any at all for 25 page markdown notes as long as Highlightr is disabled (see [1] in my previous comment).

**Comment by quinncomendant** [comment]
💯  agree slow loading is better than glitchy loading. But I don't see why we need to compromise. I use lots of other text editors which can open multi-megabyte text files in less than 1 second. Are some people using huge notes non contiguous layout would provide some advantage?

Ok, I just did a quick test: I copied a text file of 7MB to my notes folder, then timed opening it in FSNotes. It took 3 seconds to load it. Other text editors open the same file instantaneously. So there is room for improving performance, but that delay is still better than having the file load in chunks.

**Comment by pdanford** [comment]
Yeah, I can see your point. But to be fair, 7MB seems like it should be an edge case regarding size. It's FSNotes, not FSBooks :) That is to say, I don't think FSNotes was intended to be a full text editor - to wit, there's a shortcut to launch a note in an external editor of your choice. For its internal editor, it uses Apple's NSTextStorage to handle the editing/loading I believe, which isn't in the same class as say Sublime Text. It would be a lot of work to implement one to be comparable to that.

And the real issue IMHO is not delay vs having the file load in chunks. It's that the option to load in chunks causes weird editing behavior. See the video in issue #1082. And to reiterate, FSNotes gives me a great experience with markdown notes of 25 pages (I'm not seeing any load delays) as long as I do [1] in my above comment. Oh, and I do [1] instead of un-checking the "Code block live highlighting (app restart required)" preference because FSNotes does do code fence block background highlighting well/fast, which I like having.

PS: For your 3 second load result test, was it markdown and was "Code block live highlighting (app restart required)" selected? Because that option will cause serious loading delay even for 45 kB markdown files with code.

**Comment by gingerbeardman** [comment]
It was my understanding that all text rendering, in editing and preview mode is done in a web view. What's your take @glushchenko?

**Comment by pdanford** [comment]
Jumping line, maybe you're thinking of the preview mode accessed with the <kbd>command</kbd><kbd>`</kbd> which is a web rendering (i.e. the pretty version). The edit mode being talked about here is instead part of Apple's TextKit. Please correct if wrong though :)

**Comment by gingerbeardman** [comment]
@pdanford no worries at all.

The Markdown view is rendering using [Down](https://github.com/johnxnguyen/Down)'s DownView that generates a Web View. In turn the Syntax Highlighting, MathJax, Mermaid charts and image loading are all handled through JavaScript: https://github.com/glushchenko/fsnotes/tree/master/Resources/DownView.bundle/js

So I think the issue here is that the Web View JavaScript performance is suboptimal and we're trying to make changes out side of that which aren't giving any useful improvements.

I see that highlight.js needs upgrading to the latest version, which could help but I wouldn't be expecting huge changes.

The only real way to solve this would be to bring the text rendering, highlighting etc. into native code. But that's a huge undertaking.

**Comment by pdanford** [comment]
Right. But to clarify, do you mean markdown view is rendered by DownView while in edit mode too?

Looking though the code, I didn't think DownView is part of the edit mode - and the edit jumping/chunk load issue isn't related to that - only to the non contiguous mode in TextKit. DownView would only come into the picture when you switch to the Preview (i.e. pretty non-edit mode), right?

Thanks!

PS: Looks like there's two .js syntax highlighters at work here. The one you mentioned in the DownView bundle (used in the pretty rendered presentation) and Highlightr which is the one used in edit mode (that's the one that I disable for performance - which is separate from the TextKit edit issue). I still have full syntax highlighting in the rendered preview mode. Here's their paths in the app:

/Applications/FSNotes.app/Contents/Resources/DownView.bundle/js/highlight.min.js
and
/Applications/FSNotes.app/Contents/Frameworks/Highlightr.framework/Versions/A/Resources/highlight.min.js

**Comment by gingerbeardman** [comment]
I think you're right about the structure.

I also see that:

> Highlightr relies on iOS & macOS JavaScriptCore to parse the code using highlight.js

So again we are back to JavaScript.

**Comment by glushchenko** [comment]
Hi, guys!

Preview mode is WKWebView.  Markdown renders to html and preview works with JS libs, like highlight.js, mathjax.js etc.

Edit mode is NSTextView, highlighting written on Swift, regexp and custom rules. Only for code blocks highlighting uses JavaScriptCore and highlight.js.

Code highlighting should of course be rewritten to Swift, but I don't have time for that now.

**Comment by glushchenko** [comment]
## Edit mode

1) Option removed completely and disabled. At this moment only one proper way is allowsNonContiguousLayout = false.
2) Added new option to improve performance with code block highlighting.  You can disable syntax highlighter, but code blocks will still have background. By default syntax highlighter enabled.

```
defaults write co.fluder.FSNotes codeBlocksWithSyntaxHighlighting -bool false 
```

## Preview mode

No changes

**Comment by glushchenko** [comment]
https://github.com/glushchenko/fsnotes/releases/tag/4.9.11

**Comment by pdanford** [comment]
Very cool! Thank you!

**Comment by glushchenko** [comment]
Guys, we have degradation from disabled "Non Contiguous Layout"  here:

https://github.com/glushchenko/fsnotes/issues/1097

Tryed enable this option, and seems now enabled "Non Contiguous Layout" works fine.

But i know about some limitations, if you want disable option and rever previus behaviur, you can do it:

`defaults write co.fluder.FSNotes allowsNonContiguousLayout -bool false`

**Comment by glushchenko** [comment]
Guys, we have degradation from disabled "Non Contiguous Layout"  here:

https://github.com/glushchenko/fsnotes/issues/1288

Tryed enable this option, and seems now enabled "Non Contiguous Layout" works fine.

But i know about some limitations, if you want disable option and rever previus behaviur, you can do it:

`defaults write co.fluder.FSNotes allowsNonContiguousLayout -bool false`
